File: src/algorithms/arrays/sliding-window-max-deque/sources/sliding-window-max-deque.py
# Sliding Window Maximum (Deque) — O(n) monotonic decreasing deque
from collections import deque as Deque

def sliding_window_max_deque(input_array: list[int], window_size: int) -> list[int]:
    array_length = len(input_array)
    if array_length == 0 or window_size <= 0 or window_size > array_length:  # @step:initialize
        return []  # @step:initialize

    result = []  # @step:initialize
    index_deque = Deque()  # @step:initialize — stores indices, front = max of current window

    for current_index in range(array_length):
        # Remove indices outside the current window from the front
        while index_deque and index_deque[0] < current_index - window_size + 1:  # @step:compare
            index_deque.popleft()  # @step:visit

        # Remove indices of elements smaller than the current element from the back
        while index_deque and input_array[index_deque[-1]] < input_array[current_index]:  # @step:compare
            index_deque.pop()  # @step:visit

        index_deque.append(current_index)  # @step:visit

        # The window is fully formed once current_index >= window_size - 1
        if current_index >= window_size - 1:  # @step:compare
            result.append(input_array[index_deque[0]])  # @step:visit

    return result  # @step:complete

File: src/algorithms/arrays/sliding-window-max-deque/sources/sliding-window-max-deque.ts
// Sliding Window Maximum (Deque) — O(n) monotonic decreasing deque
function slidingWindowMaxDeque(inputArray: number[], windowSize: number): number[] {
  const arrayLength = inputArray.length;
  if (arrayLength === 0 || windowSize <= 0 || windowSize > arrayLength) {
    // @step:initialize
    return []; // @step:initialize
  }

  const result: number[] = []; // @step:initialize
  const deque: number[] = []; // @step:initialize — stores indices, front = max of current window

  for (let currentIndex = 0; currentIndex < arrayLength; currentIndex++) {
    // Remove indices outside the current window from the front
    while (deque.length > 0 && deque[0]! < currentIndex - windowSize + 1) {
      // @step:compare
      deque.shift(); // @step:visit
    }

    // Remove indices of elements smaller than the current element from the back
    while (deque.length > 0 && inputArray[deque[deque.length - 1]!]! < inputArray[currentIndex]!) {
      // @step:compare
      deque.pop(); // @step:visit
    }

    deque.push(currentIndex); // @step:visit

    // The window is fully formed once currentIndex >= windowSize - 1
    if (currentIndex >= windowSize - 1) {
      // @step:compare
      result.push(inputArray[deque[0]!]!); // @step:visit
    }
  }

  return result; // @step:complete
}
